java/java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToLambdaPreconditionChecker.java

```diff
@@ -137,9 +137,48 @@ private MethodTree getMethodFromFunctionalInterface(TreePath pathToNewClassTree)
                 candidate = (MethodTree)member;
             }
         }
+        // default methods can't be implemented with a lambda
+        ExecutableElement candidateElement = (ExecutableElement) info.getTrees().getElement(new TreePath(pathToNewClassTree, candidate));
+        if (overridesDefaultMethod(candidateElement, (TypeElement) baseElement)) {
+            return null;
+        }
         return candidate;
     }
 
+    private boolean overridesDefaultMethod(ExecutableElement method, TypeElement superType) {
+        Boolean overrides = overridesDefaultMethodImpl(method, superType);
+        return overrides != null && overrides;
+    }
+
+    private Boolean overridesDefaultMethodImpl(ExecutableElement method, TypeElement superType) {
+        List<? extends VariableElement> methodParams = method.getParameters();
+        for (Element e : superType.getEnclosedElements()) {
+            if (e.getKind() == ElementKind.METHOD && method.getSimpleName().equals(e.getSimpleName())
+                    && parameterTypesMatch(methodParams, ((ExecutableElement) e).getParameters())) {
+                return e.getModifiers().contains(Modifier.DEFAULT);
+            }
+        }
+        for (TypeMirror otherType : superType.getInterfaces()) {
+            Boolean overrides = overridesDefaultMethodImpl(method, (TypeElement) types.asElement(otherType));
+            if (overrides != null) {
+                return overrides;
+            }
+        }
+        return null; // no match here but check the rest of the interface tree
+    }
```

Comment:
I was a bit worried that this could loop forever if the sourcecode has an (invalid) cyclic interface hierarchy. But it looks like the compiler quits in time so that cycles don't happen at this stage.

Adding a HashSet and checking for endless recursions would be easy but appears to be not needed.

Reply:
@lahodaj javac won't create the model of invalid code which produces cycles in interface hierarchies, right? I tested it and this seems to work fine.